ccf_specs.md

```diff
@@ -161,6 +161,74 @@ This specification uses CDDL notation to express CBOR data items:
+
+Decoders MUST reject CCF encodings that are not valid unless otherwise specified in this section.
+
+A CCF encoding complies with "Valid CCF Encoding Requirements" if it complies with the following restrictions:
+
+- CBOR data items MUST be well-formed and valid as defined in RFC 8949.  E.g. CBOR text strings MUST contain valid UTF-8.  As an exception, RFC 8949 requirements for CBOR maps are not applicable because CCF does not use CBOR maps.
+
+- `composite-type.id` MUST be unique in `ccf-typedef-message` or `ccf-typedef-and-value-message`.
+
+- `composite-type.cadence-type-id` MUST be unique in `ccf-typedef-message` or `ccf-typedef-and-value-message`.
+
+- `type-ref.id` MUST refer to `composite-type.id`.
+
+- `composite-type-value.id` MUST be unique in the same `composite-type-value` data item.
+
+- `type-value-ref.id` MUST refer to `composite-type-value.id` in the same `composite-type-value` data item.
+
+- `name` MUST be unique in `composite-type-value.fields`.
+
+- `identifier` MUST be unique in `composite-type-value.initializers`.
+
+- `restricted-type.restrictions` MUST be unique.
+
+- `field-name` MUST be unique in `composite-type`.
+
+- keys MUST be unique in `dict-value`.  Decoders are not always required to check for duplicate dictionary keys.  In some cases, checking for duplicate dictionary key is not necessary or it may be delegated to the application.
+
+### Deterministic CCF Encoding Requirements
+
+A CCF encoding is deterministic if it satisfies the "Deterministic CCF Encoding Requirements".
+
+Encoders SHOULD emit CCF encodings that are deterministic.  CCF-based protocols MUST specify when encoders are required to emit deterministic CCF encodings.
+
+Decoders SHOULD check CCF encodings to determine whether they are deterministic encodings.  CCF-based protocols MUST specify when decoders are required to check for deterministic encodings and how to handle nondeterministic encodings.
+
+A CCF encoding satisfies the "Deterministic CCF Encoding Requirements" if it satisfies the following restrictions:
+
+- CCF encodings MUST satisfy "Valid CCF Encoding Requirements" defined in this document.
+
+- CCF encodings MUST satisfy "Core Deterministic Encoding Requirements" defined in RFC 8948 Section 4.2.1.  As an exception, RFC 8949 requirements for CBOR maps are not applicable because CCF does not use CBOR maps.
+
+- `composite-type.id` in `ccf-typedef-and-value-message` MUST be identical to its zero-based index in `composite-typedef`.
+
+- `composite-type-value.id` MUST be identical to the zero-based encoding order `type-value`.
+
+- `inline-type-and-value` MUST NOT be used when type can be deduced, for instance:
+  - array elements MUST use `inline-type-and-value` if array type is `[AnyStruct]`.
```

Comment:
Likewise, maybe the same for `[AnyResource]` (the other kind of top-level type)?

Reply:
Thanks for bringing this up before today's meeting!  :+1:  

With commit https://github.com/fxamacker/ccf_draft/pull/57/commits/d36c400e4f13e56933755527d6c5611c5fd3f866 , I added a new section, "Cadence Types and Values Encoding" under "Serialization Considerations" to clarify.

Summary of changes in the commit:

```
Under "Serialization Considerations", create a section to clarify encoding of Cadence types and values.

- Mention that for compactness, CCF encoding skips Cadence type encoding when type can be inferred.

- Describe CCF encoding when type can be inferred by using array of String to illustrate.

- Describe CCF encoding when type cannot be inferred by using array of AnyStruct to illustrate.

- Remove 4 bullet points in "Deterministic CCF Encoding Requirements" because they are replaced by
the new section, "Cadence Types and Values Encoding".
```